Implement hot reload using the DDC library bundle format

srujzs opened the pull request:
https://github.com/dart-lang/webdev/issues/2516

- Updates restart/reload code to accept a resetCompiler boolean to disambiguate between whether this is a full restart and whether to reset the resident compiler.
- Adds code to call reloadSources in DWDS and handle the response (including any errors).
- Adds code to invoke reassemble.
- Adds code to emit a script that DWDS can later consume that contains the changed sources and their associated libraries. This is used to hot reload. The bootstrapper puts this in the global window. DWDS should be updated to accept it in the provider itself. See https://github.com/dart-lang/webdev/issues/2584.
- Adds code to parse module metadata from the frontend server. This is identical to the implementation in DWDS % addressing type-related lints.
- Adds tests that run the existing hot reload tests but with web. Some modifications are mode, including waiting for Flutter runs to finish executing, and skipping a test that's not possible on the web.

Needs DWDS 24.3.4 to be published first and used before we can land.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

srujzs committed:
Implement hot reload for the web

- Updates restart/reload code to accept a resetCompiler boolean
to disambiguate between whether this is a full restart and whether
to reset the resident compiler.
- Adds code to call reloadSources in DWDS and handle the response
(including any errors).
- Adds code to invoke reassemble.
- Adds code to emit a script that DWDS can later consume that contains
the changed sources and their associated libraries. This is used to
hot reload. The bootstrapper puts this in the global window. DWDS should
be updated to accept it in the provider itself. See
https://github.com/dart-lang/webdev/issues/2584.
- Adds code to parse module metadata from the frontend server. This is
identical to the implementation in DWDS % addressing type-related lints.
- Adds tests that run the existing hot reload tests but with web. Some
modifications are mode, including waiting for Flutter runs to finish
executing, and skipping a test that's not possible on the web.

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotreload2

srujzs committed:
Update comments

nshahan commented on packages/flutter_tools/lib/src/isolated/devfs_web.dart:
```diff
@@ -179,6 +180,29 @@ class WebAssetServer implements AssetReader {
   }
 
+  /// Given a list of [modules] that need to be reloaded, writes a file that
+  /// contains a list of objects each with two fields:
```
Personally I find it helpful to just see an example of the json output in the comment.

srujzs replied:
Added a json code block here and for `performRestart`.

nshahan commented on packages/flutter_tools/lib/src/devfs.dart:
```diff
@@ -566,6 +566,7 @@ class DevFS {
     AssetBundle? bundle,
     bool bundleFirstUpload = false,
     bool fullRestart = false,
+    bool resetCompiler = false,
```
I see that this API has no docs for all these arguments but since you are adding one now, does `resetCompiler` mean send the reset signal to the frontend server?

Also what does `fullRestart` mean now?

srujzs replied:
Indeed, `fullRestart` means the same as before - we were just using it to also signal whether we do the reset. I know the parameters aren't documented so it looks weird, but I think adding docs for the parameters I'm adding/modifying makes sense.

nshahan commented on packages/flutter_tools/lib/src/isolated/devfs_web.dart:
```diff
@@ -179,6 +180,29 @@ class WebAssetServer implements AssetReader {
+      final List<String> libraries = metadata.libraries.keys.toList();
+      moduleToLibrary.add(<String, Object>{'src': module, 'libraries': libraries});
+    }
+    writeFile('main.dart.js.reloadScripts', json.encode(moduleToLibrary));
```
Is the "main.dart" portion of this filename based on the hard coded name of a wrapping script? If it is always the same, would something like "reload_scripts.json" make more sense? I also try to keep camel case out of filenames in the dart-y world.

srujzs replied:
So initially I wanted to emulate what we did internally for hot restart by adding a query suffix, but then I was worried that the cache busting wouldn't play nicely with that so I ended up with this.

I agree this would be much nicer just writing the restart and reload scripts like you propose and reading that path later. Done.

bkonyi commented on packages/flutter_tools/lib/src/web/module_metadata.dart:
```diff
@@ -0,0 +1,189 @@
+class LibraryMetadata {
+  LibraryMetadata(this.name, this.importUri, this.partUris);
+
+  LibraryMetadata.fromJson(Map<String, dynamic> json)
```
Here and elsewhere: we should replace `dynamic` with `Object?`

srujzs replied:
Done.

nshahan commented on packages/flutter_tools/lib/src/web/module_metadata.dart:
```diff
@@ -0,0 +1,189 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
This file is just a duplicate copy, no modifications right? Maybe we should leave some kind of link in the comment to the original source of truth?

srujzs replied:
Mostly. It required a couple tweaks to make type check-related lints happy within this repo.

Done.

bkonyi commented on packages/flutter_tools/test/integration.shard/hot_reload_test.dart:
```diff
@@ -16,6 +16,10 @@ import 'test_driver.dart';
 import 'test_utils.dart';
 
 void main() {
+  testAll();
```
I think it'd be better if we kept the test logic in a separate file (e.g., `hot_reload_test_common.dart`) and have the runtime specific test files both import the file.

Also, I'd wrap each invocation of `testAll()` with a `group` that has a name that includes the runtime target so that it'll be more clear which version of the test is running.

srujzs replied:
> have the runtime specific test files both import the file

The Flutter analyze bots gave me a friendly reminder to do so as well. :D I'll have to move this to `test_data` to avoid a separate analysis issue: https://github.com/flutter/flutter/blob/385878e7c1448df4762a7a7f99a662bb3c5d2ed4/dev/bots/analyze.dart#L561.

Done, and modified the restart tests to be consistent here as well.

nshahan reviewed (approved):
Thanks so much for getting the existing tests running on the web too!

bkonyi commented on packages/flutter_tools/lib/src/web/module_metadata.dart:
```diff
@@ -0,0 +1,189 @@
+  LibraryMetadata(this.name, this.importUri, this.partUris);
+
+  LibraryMetadata.fromJson(Map<String, dynamic> json)
+    : name = _readRequiredField(json, 'name'),
```
We should have constants defined for keys that are referenced in multiple places.

srujzs replied:
Done.

bkonyi commented on packages/flutter_tools/lib/src/web/module_metadata.dart:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  ModuleMetadata.fromJson(Map<String, dynamic> json)
+    : version = _readRequiredField(json, 'version'),
```
Ditto

srujzs replied:
Done.

srujzs committed:
Use DWDS 24.3.4

srujzs committed:
Address review comments

- Move tests to shared common test locations and add a group to
make it clear whether we're running on the web or not.
- Remove dynamic in favor of Object? in module metadata.
- Add constants for field names in module metadata.
- Add documentation for json files and parameters.
- Change named of restart and reload scripts.

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotreload2

srujzs commented:
> Thanks so much for getting the existing tests running on the web too!

There's one I have yet to add (`hot_reload_errors_test`) but that'll come when we get the `recompile-request` change in.

srujzs committed:
Remove accidental flutter.resume in test

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotreload2

srujzs committed:
Move awaiting for stateless in test up before the hot reload

srujzs committed:
Add spaces for formatting

srujzs committed:
Move run to after subscription registration

srujzs committed:
Merge remote-tracking branch 'upstream/main' into hotreload2

srujzs committed:
Testing logs for windows bots

srujzs committed:
Add print statements v2

srujzs committed:
remove print statements for debugging

srujzs committed:
Skip windows tests for web due to run failures

srujzs committed:
Add intended skip markers

srujzs committed:
Add skips to tests instead of common
